src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,178 @@
+        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            var closingPoint = braceCompletionContext.ClosingPoint;
+            if (closingPoint < 1)
+            {
+                return null;
+            }
+
+            var openingPoint = braceCompletionContext.OpeningPoint;
+            var document = braceCompletionContext.Document;
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (sourceText[openingPoint] != OpeningBrace)
+            {
+                return null;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(openingPoint, findInsideTrivia: true);
+            var validOpeningPoint = await IsValidOpenBraceTokenAtPositionAsync(token, openingPoint, document, cancellationToken).ConfigureAwait(false);
+            if (!validOpeningPoint)
+            {
+                return null;
+            }
+
+            var braceTextEdit = new TextChange(TextSpan.FromBounds(closingPoint, closingPoint), ClosingBrace.ToString());
+            // The caret location should be right before where the closing brace was inserted.
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation: openingPoint + 1);
+        }
+
+        ///<inheritdoc cref="IBraceCompletionService.GetTextChangesAfterCompletionAsync(BraceCompletionContext, CancellationToken)"/>
+        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        ///<inheritdoc cref="IBraceCompletionService.GetTextChangeAfterReturnAsync(BraceCompletionContext, bool, CancellationToken)"/>
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, bool supportsVirtualSpace, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        ///<inheritdoc cref="IBraceCompletionService.IsValidForBraceCompletionAsync(char, int, Document, CancellationToken)"/>
+        public virtual async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            // check that the user is not typing in a string literal or comment
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```

Comment:
can you check the OpeningBrace thing first, then actually do the exensive stuff like getting the tree?